### PR TITLE
Frozen string literal

### PR DESCRIPTION
**Pull request by rrosenblum**
Now that Ruby 2.3.0 has an official version, we can handle the work needed for frozen string literals. @sachin21 started this work in #2328. There are many more strings that will need to be rewritten. 

I started the work of fixing the frozen string literal errors, but it will take some time and digging to fix them all.

### CODE REVIEW

**Comment by sachin21** [comment]
Awesome! +1

**Comment by rrosenblum** [comment]
@bbatsov, the build is passing as is. I currently have Ruby 2.3.0 as allowed failures in Travis CI because of the frozen string literal issues. Should the frozen string literal issues be fixed as a part of this PR or a separate PR?

**Comment by bbatsov** [comment]
I'd say the issues should be fixed in this PR.

**Comment by bquorning** [comment]
Perhaps you could set `# frozen_string_literal: false` instead of `true` where needed and get a green build?

**Comment by rrosenblum** [comment]
> Perhaps you could set # frozen_string_literal: false instead of true where needed and get a green build?

That is what I am trying for the time being.

**Comment by bbatsov** [comment]
Btw, when we're ready with this PR we should add a cop similar to the Encoding cop that will check if files are using frozen strings and add the magic comment automatically if they're not. I think many people would find this useful with the upcoming transition to Ruby 3.0.

**Comment by rrosenblum** [comment]
I think that is a great idea. I think the configuration would be safe to check for `freeze` and `<<` being called on string literals.

**Comment by bbatsov** [comment]
How are things going here? I see there's some failure on JRuby 9k.

**Comment by rrosenblum** [comment]
I am trying to figure out the failure in JRuby 9k. I have the cop functioning as I want it to. There are still some failures trying to run the code after enabling the `frozen_string_literal` comment. I am still trying to find the best way to handle those.

**Comment by rrosenblum** [comment]
The JRuby 9k failure is a non issue now.

**Comment by sachin21** [comment]
Good job!! Thank you for resolving PR in place of me!

**Comment by bbatsov** [comment]
The new cop should make use of the functionality @alexdowad just added about target Ruby versions.

**Comment by rrosenblum** [comment]
I added functionality to target Ruby >= 2.3.0 when configured with `when_needed` let me know if you think this should apply to `always` as well. 

When I change the configuration to `always`, and enable frozen string literals on all files, I am getting some issues that will take some digging to find a solution to. The error is:
```
rubocop/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.4.1/lib/rspec/core/source.rb:23:in `split': invalid byte sequence in US-ASCII (ArgumentError)
```
I tried line 32 from ProcessedSource, `source.force_encoding(Encoding::UTF_8)`, it helps fix a large number of errors, but the above error still persists.

**Comment by alexdowad** [comment]
@rrosenblum I've seen that `split` error a number of times. I have never found the root cause, but what I have found is: it's trying to print a stack trace which was caused by another exception, and when it is printing the stack trace, something causes that error. So it's a symptom of another uncaught exception. You need to find what that "other" uncaught exception is.

**Comment by rrosenblum** [comment]
Thanks, I was curious why I didn't see more exceptions being printed. I unfortunately may need to work through some of these files one by one to try and find it. Pry doesn't seem to be helping much either.

**Comment by rrosenblum** [comment]
@alexdowad I am having trouble using the new configuration for `TargetRubyVersion`. I added the configuration option to the `default.yml`, and I am using `context 'ruby < 2.3', :ruby19 do`, but the tests still run when using Ruby 2.3.0. Can you take a look and see what I am doing wrong?

**Comment by alexdowad** [comment]
Hi @rrosenblum. Just looking at it right now.

**Comment by alexdowad** [comment]
@rrosenblum please pull the `help_rrosenblum` branch from alexdowad/rubocop. Your specs are passing.

**Comment by alexdowad** [comment]
OK, I found a few more things:

- The problem with RSpec is coming from `rspec-core/lib/rspec/core/source.rb:13`. Change `File.read` to `File.binread`, and then you can see the actual error message from RuboCop.
- Your addition of `if RUBY_VERSION > ...` in `ProcessedSource#parse` breaks one of the processed source specs. I recommend you just do a `git rebase -i` and drop that commit.
- You have configured the new cop as enabled in `.rubocop.yml`, but the cop contains a validation method which raises an exception if `TargetRubyVersion` is lower than 2.3. Is this cop really not intended to be used on any project which supports any Ruby version lower than 2.3?

**Comment by rrosenblum** [comment]
Thanks for the help @alexdowad. I will take a look at your changes. I still have some work to do on this cop. 

> The problem with RSpec is coming from rspec-core/lib/rspec/core/source.rb:13. Change File.read to File.binread, and then you can see the actual error message from RuboCop.

Do you think this should be checked into the code base?

> Your addition of if RUBY_VERSION > ... in ProcessedSource#parse breaks one of the processed source specs. I recommend you just do a git rebase -i and drop that commit.

I have some work to do on this. Removing the `RUBY_VERSION >` check causes nearly all of the tests to fail. There is an issue calling `force_encoding` on a string in Ruby 2.3. Ruby considers this modifying the frozen string. According to the comment, `force_encoding` should only be needed for Ruby 1.9.3 since encoding defaults to `UTF-8` in Ruby 2.0+. I am going to continue digging into this. I think I can have the test target a Ruby version and have the check in `ProcessedSource#parse` check the encoding of the string instead of the `RUBY_VERSION`.

> You have configured the new cop as enabled in .rubocop.yml, but the cop contains a validation method which raises an exception if TargetRubyVersion is lower than 2.3. Is this cop really not intended to be used on any project which supports any Ruby version lower than 2.3?

I hadn't fully thought through that scenario. I was intending for the cop to not register an offense when running on Ruby < 2.3. I had pulled in the validation when I was trying to get `TargetRubyVersion` to work. I am not fully sure how we want to handle this. The cop would be safe to run against Ruby < 2.3, but users will not be able to see if the comment works or not.

**Comment by rrosenblum** [comment]
Part of my issue was that `ruby_greater_tor_equal` and `ruby_less_than` were removed from the tests. It doesn't seem like there is a replacement for these. I was trying to use `target_ruby_version` to accomplish something that it was not designed to do.

**Comment by alexdowad** [comment]
> Do you think this should be checked into the code base?

I do think it should be checked into the RSpec code base, but I don't have commit rights for RSpec. :smile:

> There is an issue calling force_encoding on a string in Ruby 2.3. Ruby considers this modifying the frozen string. According to the comment, force_encoding should only be needed for Ruby 1.9.3 since encoding defaults to UTF-8 in Ruby 2.0+

Aha.

Well, it may be that the spec in `processed_source_spec.rb:104` (that's the one which was failing) is testing something we don't actually need to be worried about. Maybe that test can be deleted?

All source code which goes into `ProcessedSource` will either be 1) read from a file or 2) built from string literals in a RC source file.

RC's source is all UTF-8, so any code we build from string literals should be UTF-8. In `ProcessedSource.from_file:17`, it might be better to provide a specific internal encoding. Then we could ensure that *all* code going into `ProcessedSource` would *always* be UTF-8. Then there wouldn't be any need to use `force_encoding`.

The other option is to use a non-destructive method to transcode strings to UTF-8 when needed.

> I was intending for the cop to not register an offense when running on Ruby < 2.3.

The Ruby version which an inspected project targets is now (intentionally) decoupled from the Ruby which RC itself is running on. So it would be better if the cop doesn't change its behavior depending on `RUBY_VERSION`. `.rubocop.yml` should completely describe a project's desired coding style, which leaves no need to consult `RUBY_VERSION`.

`.rubocop.yml` can be checked into source control, diffed, etc. The Ruby binary which happens to be running on my PC right now can't. `.rubocop.yml` can be easily kept in sync between different developers and different environments (including CI). Ruby interpreter binaries can't.

When (like me) you have a dozen different Ruby interpreters installed on your PC, you can easily forget which one you are using at the moment. One more reason why changing behavior depending on `RUBY_VERSION` is a poor idea.

> The cop would be safe to run against Ruby < 2.3, but users will not be able to see if the comment works or not.

Well, you should only include that `validate_config` method if the cop is never intended to be used on < 2.3 projects, and if enabling the cop on a < 2.3 project indicates faulty configuration. I'm not sure if that is the case here.

**Comment by rrosenblum** [comment]
I follow the reason for the change from using `RUBY_VERSION` to having a `TargetedRubyVersion`. In this case, I think we are going to need to use `RUBY_VERSION`. All code that we write to accommodate `frozen_string_literal` is going to be backwards compatible. The effects of adding the comment can only be seen when running Ruby 2.3.0+. I will expand the documentation to talk about this. My thought now is that `always` should always add the comment. `when_needed` will only add the comment when running Ruby 2.3.0+ and an offense is found. 

Even though this is mainly intended for Ruby 2.3.0+ it can/still will be used in lower versions. It may make writing new code a bit annoying if the code is developed using Ruby < 2.3.0 since failures will not be seen locally. Luckily, for RuboCop, all errors will be caught in a PR when Travis CI runs.

**Comment by alexdowad** [comment]
It's your code, but I still think consulting `RUBY_VERSION` is a mistake. Ruby on dev machine != Ruby in production environment. `TargetRubyVersion` tells RC what Ruby will be used in production (whether that is a server, or an end-user's machine).

Your call.

**Comment by rrosenblum** [comment]
For all other cops, I agree with you. This cop is an exception to the rule. For all other cops, it has been make sure that you do not use a new feature so that the code is compatible with a lower Ruby version. This is the first cop that is adding in a feature to be compatible with the future. It will help people to upgrade their code base.

**Comment by rrosenblum** [comment]
All tests are passing. @bbatsov this is ready for review.

**Comment by rrosenblum** [comment]
@bbatsov and @alexdowad what do you think about this cop moving from `disabled` to `enabled` configured for `when_needed`?

**Comment by alexdowad** [comment]
> what do you think about this cop moving from disabled to enabled

Hmm. The purpose of this cop is to help people ready their code to run on Ruby 3.0.

Do you think most (or many) of our users are interested in preparing for Ruby 3.0 right now?

**Comment by rrosenblum** [comment]
I forgot to run RuboCop after the rebase, I will take care of the error as soon as I get a chance.

> Do you think most (or many) of our users are interested in preparing for Ruby 3.0 right now?

Good point. We can always move it to enabled once Ruby 3.0 is closer to release.

**Comment by bbatsov** [comment]
Well, even now the cop is pretty useful (if you're on 2.3) - no more annoying `.freeze`s all over your codebase. I'm leaning towards enabling it by default if the target Ruby version is 2.3 (or newer).

**Comment by rrosenblum** [comment]
I am fine with either option. I can understand waiting until Ruby 3 is closer, but I agree with you @bbatsov, I think that it adds value to any code base that is running Ruby 2.3. It is going to make upgrading down the line easier, and if the code does not need to be backwards compatible, then `freeze` can be removed. 

I wasn't thinking about removing calls to `freeze` until you mentioned something. We can modify  `RedundantFreeze` to also check for calls to `freeze` on string literals if the `TargetRubyVersion` is 2.3+.

**Comment by rrosenblum** [comment]
I changed the cop from disabled to enabled. I also made a tweak to the `when_needed` config to account for an edge case where the user is using Ruby < 2.3 and the `TargetRubyVersion` is 2.3+.

**Comment by bquorning** [comment]
I stand by my earlier commented that

> perhaps you could set `# frozen_string_literal: false` instead of `true` where needed and get a green build?

The discussion in this PR has understandably been mostly about the cop itself, but the last commit changes a whole lot of code that it might make sense to discuss in a separate PR.

**Comment by rrosenblum** [comment]
@bquorning I initially did what you mentioned. While I was going through looking at offenses, nothing seemed too difficult to fix. I can understand where this is a large PR and some changes may get lost, but I am not sure that 1 PR vs 2 is going to matter much. I can disable frozen string literals in the files from the commit `Fix frozen string literal offenses` and add it to separate PR if that would make it easier. @bbatsov what are your thoughts on 1 PR vs 2?

**Comment by bbatsov** [comment]
Smaller changes are easier to review. :-)

**Comment by rrosenblum** [comment]
I have made the change to disable frozen string literals in this PR. I have created PR #2585 to fix the errors.

**Comment by rrosenblum** [comment]
@bbatsov this is ready for code review. If you would like me to make this PR smaller, I could leave to cop disabled in this PR and open another PR that adds the comment to all files. I would also have the fix frozen string literals PR for a total of 3 PRs for this work.

**Comment by bbatsov** [comment]
The tests are failing, otherwise the code looks ok.

**Comment by rrosenblum** [comment]
Whoops, I am not used to rebase braking so much. I have fixed the failures.

**Comment by sachin21** [comment]
:tada: :tada: :tada:

File: config/enabled.yml [hunk 1]
```diff
@@ -265,6 +265,12 @@ Style/FileName:
   StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#snake-case-files'
   Enabled: true
 
+Style/FrozenStringLiteralComment:
+  Description: >-
+                 Add the frozen_string_literal comment to the top of files
+                 to help transition from Ruby 2.3.0 to Ruby 3.0.
+  Enabled: false
```
Comment (bquorning) [inline config/enabled.yml]: This configuration should have been in `config/disabled.yml`, right?

Comment (bbatsov) [inline config/enabled.yml]: Ops. Yes.

Comment (rrosenblum) [inline config/enabled.yml]: Whoops, it should be in `config/enabled.yml` and set to `true` as per an earlier discussion. I will correct this and submit a PR.

**Pull request merged**